std/hash/poseidon_gl.asm

```diff
@@ -58,129 +60,57 @@ machine PoseidonGL(LASTBLOCK, operation_id) {
+    let x7: expr[STATE_SIZE] = array::zip(x6, a, |x6, a| x6 * a);
 
     // Apply S-Boxes on the first element and otherwise if it is a full round.
-    pol b0 = x7_0;
-    pol b1 = PARTIAL * (a1 - x7_1) + x7_1;
-    pol b2 = PARTIAL * (a2 - x7_2) + x7_2;
-    pol b3 = PARTIAL * (a3 - x7_3) + x7_3;
-    pol b4 = PARTIAL * (a4 - x7_4) + x7_4;
-    pol b5 = PARTIAL * (a5 - x7_5) + x7_5;
-    pol b6 = PARTIAL * (a6 - x7_6) + x7_6;
-    pol b7 = PARTIAL * (a7 - x7_7) + x7_7;
-    pol b8 = PARTIAL * (a8 - x7_8) + x7_8;
-    pol b9 = PARTIAL * (a9 - x7_9) + x7_9;
-    pol b10 = PARTIAL * (a10 - x7_10) + x7_10;
-    pol b11 = PARTIAL * (a11 - x7_11) + x7_11;
+    let b: expr[STATE_SIZE] = array::new(STATE_SIZE, |i| if i == 0 {
+        x7[i]
+    } else {
+        PARTIAL * (a[i] - x7[i]) + x7[i]
+    });
+
+    // The MDS matrix
+    let M = [
+        [25, 15, 41, 16,  2, 28, 13, 13, 39, 18, 34, 20],
+        [20, 17, 15, 41, 16,  2, 28, 13, 13, 39, 18, 34],
+        [34, 20, 17, 15, 41, 16,  2, 28, 13, 13, 39, 18],
+        [18, 34, 20, 17, 15, 41, 16,  2, 28, 13, 13, 39],
+        [39, 18, 34, 20, 17, 15, 41, 16,  2, 28, 13, 13],
+        [13, 39, 18, 34, 20, 17, 15, 41, 16,  2, 28, 13],
+        [13, 13, 39, 18, 34, 20, 17, 15, 41, 16,  2, 28],
+        [28, 13, 13, 39, 18, 34, 20, 17, 15, 41, 16,  2],
+        [ 2, 28, 13, 13, 39, 18, 34, 20, 17, 15, 41, 16],
+        [16,  2, 28, 13, 13, 39, 18, 34, 20, 17, 15, 41],
+        [41, 16,  2, 28, 13, 13, 39, 18, 34, 20, 17, 15],
+        [15, 41, 16,  2, 28, 13, 13, 39, 18, 34, 20, 17]
+    ];
 
     // Multiply with MDS Matrix
-    let c = [
-        25*b0 + 15*b1 + 41*b2 + 16*b3 +  2*b4 + 28*b5 + 13*b6 + 13*b7 + 39*b8 + 18*b9 + 34*b10 + 20*b11,
-        20*b0 + 17*b1 + 15*b2 + 41*b3 + 16*b4 +  2*b5 + 28*b6 + 13*b7 + 13*b8 + 39*b9 + 18*b10 + 34*b11,
-        34*b0 + 20*b1 + 17*b2 + 15*b3 + 41*b4 + 16*b5 +  2*b6 + 28*b7 + 13*b8 + 13*b9 + 39*b10 + 18*b11,
-        18*b0 + 34*b1 + 20*b2 + 17*b3 + 15*b4 + 41*b5 + 16*b6 +  2*b7 + 28*b8 + 13*b9 + 13*b10 + 39*b11,
-        39*b0 + 18*b1 + 34*b2 + 20*b3 + 17*b4 + 15*b5 + 41*b6 + 16*b7 +  2*b8 + 28*b9 + 13*b10 + 13*b11,
-        13*b0 + 39*b1 + 18*b2 + 34*b3 + 20*b4 + 17*b5 + 15*b6 + 41*b7 + 16*b8 +  2*b9 + 28*b10 + 13*b11,
-        13*b0 + 13*b1 + 39*b2 + 18*b3 + 34*b4 + 20*b5 + 17*b6 + 15*b7 + 41*b8 + 16*b9 +  2*b10 + 28*b11,
-        28*b0 + 13*b1 + 13*b2 + 39*b3 + 18*b4 + 34*b5 + 20*b6 + 17*b7 + 15*b8 + 41*b9 + 16*b10 +  2*b11,
-         2*b0 + 28*b1 + 13*b2 + 13*b3 + 39*b4 + 18*b5 + 34*b6 + 20*b7 + 17*b8 + 15*b9 + 41*b10 + 16*b11,
-        16*b0 +  2*b1 + 28*b2 + 13*b3 + 13*b4 + 39*b5 + 18*b6 + 34*b7 + 20*b8 + 17*b9 + 15*b10 + 41*b11,
-        41*b0 + 16*b1 +  2*b2 + 28*b3 + 13*b4 + 13*b5 + 39*b6 + 18*b7 + 34*b8 + 20*b9 + 17*b10 + 15*b11,
-        15*b0 + 41*b1 + 16*b2 +  2*b3 + 28*b4 + 13*b5 + 13*b6 + 39*b7 + 18*b8 + 34*b9 + 20*b10 + 17*b11
-    ];
+    let c: expr[STATE_SIZE] = array::map(M, |M| array::sum(array::zip(M, b, |M, b| M * b)));
```

Comment:
Maybe it's a bit less confusing to use different names for `M` and its elements and the elements of the elements? Or define an intermediate function?

Reply:
I did both ;)